### PR TITLE
[AutoDiff] Handle init_enum_data_addr and inject_enum_addr for Optional

### PR DESCRIPTION
**Pull request by asavonic**
Optional's `init_enum_data_addr` and `inject_enum_addr` instructions are generated in presence of non-loadable Optional values. The compiler used to treat these instructions as inactive, and this resulted in silent run-time issues described in #64223.

The patch marks `init_enum_data_addr` as "active" if its Optional operand is also active, and in PullbackCloner we differentiate through it and the related `inject_enum_addr`.

However, we only determine this relation in simple cases when both instructions are in the same block. There is no def-use relation between them (both take the same Optional operand), so if there is more than one set of instructions operating on the same Optional, or there is some control flow, we currently bail out.

In PullbackCloner, we walk over instructions in reverse order and start from `inject_enum_addr` and its `Optional<Wrapped>.TangentVector` operand. Assuming that is is already initialized, we emit an `unchecked_take_enum_data_addr` and set it as the adjoint buffer of `init_enum_data_addr`. The Optional value is invalidated, and we have to destroy the enum data address later when we reach `init_enum_data_addr`.

Before this patch, `Optional<Wrapped>.TangentVector` was implemented as a struct with an `Optional<Wrapped.TangentVector>` field (`value`). In order to use `unchecked_take_enum_data_addr`, we have to access the underlying Optional field, but it is not possible, because the struct is resilient:

    SIL verification failed: cannot access storage of resilient struct:
      !checkResilience(sd, F.getModule().getSwiftModule(), F.getResilienceExpansion())
    Verifying instruction:
	 %6 = alloc_stack $Optional<τ_0_0>.TangentVector
    ->   %8 = struct_element_addr %6 : $*Optional<τ_0_0>.TangentVector, #Optional.TangentVector.value
	 %9 = unchecked_take_enum_data_addr %8 : $*Optional<τ_0_0.TangentVector>, #Optional.some!enumelt

Therefore, the patch changes `Optional<Wrapped>.TangentVector` from struct to enum with same fields as Optional (`.some` and `.none`). This way we use `unchecked_take_enum_data_addr` on the `TangentVector` directly.

### CODE REVIEW

**Comment by asl** [comment]
@swift-ci please test

File: lib/SILOptimizer/Differentiation/LinearMapInfo.cpp [hunk 1]
```diff
@@ -436,6 +436,18 @@ bool LinearMapInfo::shouldDifferentiateApplySite(FullApplySite applySite) {
   return hasActiveResults && hasActiveArguments;
 }
 
+static bool shouldDifferentiateInjectEnumAddr(
+    InjectEnumAddrInst &inject, const DifferentiableActivityInfo &activityInfo,
```
Comment (asl) [inline lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:440]: Shouldn't `inject be const reference?

Comment (asl) [inline lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:440]:
```suggestion
    const InjectEnumAddrInst &inject, const DifferentiableActivityInfo &activityInfo,
```

File: lib/SILOptimizer/Differentiation/LinearMapInfo.cpp [hunk 2]
```diff
@@ -489,6 +501,11 @@ bool LinearMapInfo::shouldDifferentiateInstruction(SILInstruction *inst) {
         isa<DestroyValueInst>(inst) || isa<DestroyAddrInst>(inst))
       return true;
   }
+
```
Comment (asl) [inline lib/SILOptimizer/Differentiation/LinearMapInfo.cpp:504]:
```suggestion
    // Should differentiate `inject_enum_addr` if the corresponding `init_enum_addr` has an active operand.
```

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 3]
```diff
@@ -1616,6 +1616,74 @@ class PullbackCloner::Implementation final
...
   }
 
+  void visitInjectEnumAddrInst(InjectEnumAddrInst *inject) {
+    SILBasicBlock *bb = inject->getParent();
+    SILValue origEnum = inject->getOperand();
+    bool unsupported = false;
+
+    // Only `Optional`-typed operands are supported for now. Diagnose all other
+    // enum operand types.
+    //
```
Comment (asl) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1626]: Do we need those empty comment lines? Here and below?

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 4]
```diff
@@ -1616,6 +1616,74 @@ class PullbackCloner::Implementation final
...
+    auto *optionalEnumDecl = getASTContext().getOptionalDecl();
+    if (origEnum->getType().getEnumOrBoundGenericEnum() != optionalEnumDecl) {
+      LLVM_DEBUG(getADDebugStream()
+                 << "Unsupported enum type in PullbackCloner: " << *inject);
+      unsupported = true;
```
Comment (asl) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1631]: return missed?

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 5]
```diff
@@ -1616,6 +1616,74 @@ class PullbackCloner::Implementation final
...
+                     << *inject);
+          unsupported = true;
+          break;
+        }
+
+        origData = init;
+      }
+    }
+
+    if (unsupported) {
```
Comment (asl) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:1653]: If would fold this into two cases above. Makes more readable code

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
Tagging @BradLarson

**Comment by asl** [comment]
@asavonic Does it fix https://github.com/apple/swift/issues/64223?

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp (hunk 3 above)
Comment (asavonic) [reply lib/SILOptimizer/Differentiation/PullbackCloner.cpp]: No, not really necessary. Removed them.

**Comment by asavonic** [comment]
> @asavonic Does it fix #64223?

Yes. `optional_pullback.swift` LIT test is the reproducer from #64223.
It works fine now and gives the correct result.

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp (hunk 5 above)
Comment (asavonic) [reply lib/SILOptimizer/Differentiation/PullbackCloner.cpp]: Done.

**Comment by asavonic** [comment]
@rxwei thank you for the suggestion - attribute `@frozen` makes the struct field available to the compiler.

  A side-effect of this change is that we can now see `Optional.TangentVector` as `SILValueCategory::Object`. There is a code path for this case, but I don't think it was ever executed. Before the patch, `accumulateAdjointForOptional` did not handle this case correctly and crashed in `addToAdjointBuffer` when `originalValue` is `SILValueCategory::Object`. The patch addresses this issue as well.

  Changes from the previous revision of the patch:

  1. `Optional.TangentVector` declaration is reverted back to being a `struct`, but now it has the `@frozen` attribute to make it visible for the compiler.

  2. `unchecked_take_enum_data_addr` now operates on `Optional.TangentVector.value` field. I don't think that there is any specification that mentions this field, so we rely on it as an implementation detail of Swift stdlib.

  3. `Optional.TangentVector.value` is copied to a new alloca, because `unchecked_take_enum_data_addr` is destructive and we don't want to leave the struct partially invalidated.

  3. New function `setAdjointValueForOptional` handles `Object` adjoint values for Optional, in addition to `accumulateAdjointForOptional`, which only handles `Address` values.

  4. Changes in existing LIT tests are no longer needed.

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 6]
```diff
@@ -672,15 +672,23 @@ class PullbackCloner::Implementation final
...
+                                        SILValue wrappedAdjoint,
+                                        SILType optionalTy);
+
+  /// Accumulate SILValueCategory::Address optional value from `wrappedAdjoint`.
   void accumulateAdjointForOptional(SILBasicBlock *bb, SILValue optionalValue,
                                     SILValue wrappedAdjoint);
 
+  /// Set SILValueCategory::Object optional value from `wrappedAdjoint`.
+  void setAdjointValueForOptional(SILBasicBlock *bb, SILValue optionalValue,
+                                  SILValue wrappedAdjoint);
```
Comment (rxwei) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:690]: In AD code we use the term "adjoint value" to refer to adjoint objects, and "adjoint buffer" to refer to adjoint addresses, so I don't think there's a need to call out `SILValueCategory::Object` or `SILValueCategory::Address`.

What you could do is rename `accumulateAdjointForOptional` to make it clear at the call site, same for its arguments, for example:
```cpp
    void accumulateAdjointForOptionalBuffer(SILBasicBlock *bb, SILValue optionalBuffer,
                                            SILValue wrappedAdjointBuffer);

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp (hunk 6 above)
Comment (asavonic) [reply lib/SILOptimizer/Differentiation/PullbackCloner.cpp]: Thank you! Done.

**Review by rxwei — APPROVED** [review]

**Pull request merged**